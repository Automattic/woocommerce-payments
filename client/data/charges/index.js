/** @format */

/**
 * Internal dependencies
 */
import reducer from './reducer.gen';
import * as selectors from './selectors.gen';
import * as actions from './actions.gen';
import * as resolvers from './resolvers';

export { reducer, selectors, actions, resolvers };
export * from './hooks.gen';
