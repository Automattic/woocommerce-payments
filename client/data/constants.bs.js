// Generated by BUCKLESCRIPT, PLEASE EDIT WITH CARE


var namespace = "/wc/v3/payments";

var storeName = "wc/payments";

export {
  namespace ,
  storeName ,
  
}
/* No side effect */
