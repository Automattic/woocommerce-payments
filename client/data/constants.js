/** @format */

export const NAMESPACE = '/wc/v3/payments';
export const STORE_NAME = 'wc/payments';

export const ID_PREFIX = {
	transactions: 'transactions',
};
