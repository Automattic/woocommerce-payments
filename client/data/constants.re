let namespace = "/wc/v3/payments";
let storeName = "wc/payments";
