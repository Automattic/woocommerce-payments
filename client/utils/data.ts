/** @format */

export const getResourceId = (
	identifier: Record< string, string >
): string => {
	return JSON.stringify( identifier, Object.keys( identifier ).sort() );
};
