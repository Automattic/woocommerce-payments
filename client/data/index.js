/** @format */

/**
 * Internal dependencies
 */
import { STORE_NAME } from './constants';
import { initStore } from './store';

initStore();

export const WCPAY_STORE_NAME = STORE_NAME;

// We only ask for hooks when importing directly from 'data'.
export * from './deposits/hooks';
export * from './transactions/hooks';
export * from './charges/hooks';
export * from './timeline/hooks';
export * from './disputes/hooks';
export * from './settings/hooks';
export * from './multi-currency';
