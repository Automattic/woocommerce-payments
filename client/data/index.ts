/** @format */

/**
 * Internal dependencies
 */
import { STORE_NAME } from './constants';
import { initStore } from './store';

initStore();

// eslint-disable-next-line @typescript-eslint/naming-convention
export const WCPAY_STORE_NAME = STORE_NAME;

// We only ask for hooks when importing directly from 'wcpay/data'.
export * from './deposits/hooks';
export * from './transactions/hooks';
export * from './charges/hooks';
export * from './timeline/hooks';
export * from './disputes/hooks';
export * from './settings/hooks';
export * from './multi-currency';
export * from './card-readers/hooks';
export * from './capital/hooks';
export * from './documents/hooks';
export * from './payment-intents/hooks';
export * from './authorizations/hooks';
