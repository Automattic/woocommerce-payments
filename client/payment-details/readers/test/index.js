/** @format */
/**
 * External dependencies
 */
import * as React from 'react';
import { render, screen } from '@testing-library/react';
import { updateQueryString } from '@woocommerce/navigation';
import os from 'os';

/**
 * Internal dependencies
 */
import PaymentCardReaderChargeDetails from '../';
import { useCardReaderStats } from 'wcpay/data';
import { downloadCSVFile } from '@woocommerce/csv-export';

jest.mock( 'wcpay/data', () => ( {
	useCardReaderStats: jest.fn(),
} ) );

jest.mock( '@woocommerce/csv-export', () => {
	const actualModule = jest.requireActual( '@woocommerce/csv-export' );

	return {
		...actualModule,
		downloadCSVFile: jest.fn(),
	};
} );

const mockUseCardReaderStats = useCardReaderStats;
const mockDownloadCSVFile = downloadCSVFile;

describe( 'RenderPaymentCardReaderChargeDetails', () => {
	beforeEach( () => {
		jest.clearAllMocks();

		// the query string is preserved across tests, so we need to reset it
		updateQueryString( {}, '/', {} );

		global.wcpaySettings = {
			zeroDecimalCurrencies: [],
			connect: {
				country: 'US',
			},
			currencyData: {
				US: {
					code: 'USD',
					symbol: '$',
					symbolPosition: 'left',
					thousandSeparator: ',',
					decimalSeparator: '.',
					precision: 2,
					negativeFormat: '-',
				},
			},
		};
	} );

	test( 'renders without reader charges', () => {
		mockUseCardReaderStats.mockReturnValue( {
			readers: [],
			chargeError: false,
			isLoading: false,
		} );

		const { container } = render(
			<PaymentCardReaderChargeDetails
				chargeId={ 'mock_charge_id' }
				transactionId={ 'mock_transaction_id' }
			/>
		);
		expect( container ).toMatchSnapshot();
	} );

	test( 'renders reader charges', () => {
		const readers = [
			{
				reader_id: 1,
				count: 3,
				status: 'active',
				fee: { amount: 300, currency: 'usd' },
			},
			{
				reader_id: 2,
				count: 1,
				status: 'inactive',
				fee: { amount: 0, currency: 'usd' },
			},
		];

		mockUseCardReaderStats.mockReturnValue( {
			readers: readers,
			chargeError: false,
			isLoading: false,
		} );

		const { container } = render(
			<PaymentCardReaderChargeDetails
				chargeId={ 'mock_charge_id' }
				transactionId={ 'mock_transaction_id' }
			/>
		);
		expect( container ).toMatchSnapshot();
	} );

	test( 'renders loading state', () => {
		mockUseCardReaderStats.mockReturnValue( {
			readers: [],
			chargeError: false,
			isLoading: true,
		} );
		render(
			<PaymentCardReaderChargeDetails
				chargeId={ 'mock_charge_id' }
				transactionId={ 'mock_transaction_id' }
			/>
		);

		expect(
			screen.getByText( 'Your requested data is loading' )
		).toBeInTheDocument();
	} );

	test( 'loading error', () => {
		mockUseCardReaderStats.mockReturnValue( {
			readers: [],
			chargeError: new Error( 'test' ),
			isLoading: false,
		} );
		render(
			<PaymentCardReaderChargeDetails
				chargeId={ 'mock_charge_id' }
				transactionId={ 'mock_transaction_id' }
			/>
		);

		expect(
			screen.getByText( 'Readers details not loaded' )
		).toBeInTheDocument();
	} );

	test( 'should render expected columns in CSV when the download button is clicked', () => {
		const readers = [
			{
				reader_id: 1,
				count: 3,
				status: 'active',
				fee: { amount: 300, currency: 'usd' },
			},
		];

		mockUseCardReaderStats.mockReturnValue( {
			readers: readers,
			chargeError: false,
			isLoading: false,
		} );
		const { getByRole } = render(
			<PaymentCardReaderChargeDetails
				chargeId={ 'mock_charge_id' }
				transactionId={ 'mock_transaction_id' }
			/>
		);
		getByRole( 'button', { name: 'Download' } ).click();
		const expected = [ '"Reader id"', 'Status', 'Transactions', 'Fee' ];

		// checking if columns in CSV are rendered correctly
		expect(
			mockDownloadCSVFile.mock.calls[ 0 ][ 1 ]
				.split( '\n' )[ 0 ]
				.split( ',' )
		).toEqual( expected );
	} );

	test( 'should match the visible rows', () => {
		const readers = [
			{
				reader_id: 1,
				count: 3,
				status: 'active',
				fee: { amount: 300, currency: 'usd' },
			},
		];

		mockUseCardReaderStats.mockReturnValue( {
			readers: readers,
			chargeError: false,
			isLoading: false,
		} );
		const { getByRole, getAllByRole } = render(
			<PaymentCardReaderChargeDetails
				chargeId={ 'mock_charge_id' }
				transactionId={ 'mock_transaction_id' }
			/>
		);
		getByRole( 'button', { name: 'Download' } ).click();
		const csvContent = mockDownloadCSVFile.mock.calls[ 0 ][ 1 ];
		const csvRows = csvContent.split( os.EOL );
		const displayRows = getAllByRole( 'row' );
		expect( csvRows.length ).toEqual( displayRows.length );

		const csvFirstTransaction = csvRows[ 1 ].split( ',' );
		const displayFirstTransaction = Array.from(
			displayRows[ 1 ].querySelectorAll( 'td' )
		).map( ( td ) => td.textContent || '' );

		const displayFirstRowHead = Array.from(
			displayRows[ 1 ].querySelectorAll( 'th' )
		).map( ( th ) => th.textContent || '' );
		displayFirstTransaction.unshift( displayFirstRowHead[ 0 ] );

		expect( displayFirstTransaction[ 0 ] ).toBe( csvFirstTransaction[ 0 ] ); // reader id
		expect( displayFirstTransaction[ 1 ] ).toBe( csvFirstTransaction[ 1 ] ); // status
		expect( displayFirstTransaction[ 2 ] ).toBe( csvFirstTransaction[ 2 ] ); // count
		expect(
			parseFloat( getUnformattedAmount( displayFirstTransaction[ 3 ] ) )
		).toBe( parseFloat( csvFirstTransaction[ 3 ] ) ); // fee
	} );

	function getUnformattedAmount( formattedAmount ) {
		const amount = formattedAmount.replace( /[^0-9,.' ]/g, '' ).trim();
		return amount.replace( ',', '.' ); // Euro fix
	}
} );
