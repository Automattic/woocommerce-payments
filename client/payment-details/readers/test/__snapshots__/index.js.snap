// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderPaymentCardReaderChargeDetails renders reader charges 1`] = `
<div>
  <div
    class="wcpay-payment-details woocommerce-payments-page"
    style="max-width: 1032px;"
  >
    <div
      class="components-surface components-card woocommerce-table transactions-list woocommerce-report-table has-search has-actions has-menu css-nsno0f-View-Surface-getBorders-primary-Card-rounded em57xhy0"
      data-wp-c16t="true"
      data-wp-component="Card"
    >
      <div
        class="css-mgwsf4-View-Content em57xhy0"
      >
        <div
          class="components-flex components-card__header components-card-header css-1g5oj2q-View-Flex-sx-Base-sx-Items-ItemsRow-Header-borderRadius-borderColor-medium em57xhy0"
          data-wp-c16t="true"
          data-wp-component="CardHeader"
        >
          <h2
            class="components-truncate components-text css-1s1vggh-View-Text-sx-Base em57xhy0"
            data-wp-c16t="true"
            data-wp-component="Text"
          >
            Card readers
          </h2>
          <div
            class="woocommerce-table__actions"
          >
            <button
              class="components-button woocommerce-table__download-button"
              type="button"
            >
              <svg
                class="gridicon gridicons-cloud-download"
                height="24"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g>
                  <path
                    d="M18 9c-.009 0-.017.002-.025.003A6.496 6.496 0 005 9.5a6.5 6.5 0 00.186 1.519C5.123 11.016 5.064 11 5 11a4 4 0 00-4 4c0 1.202.541 2.267 1.38 3h18.593C22.196 17.089 23 15.643 23 14a5 5 0 00-5-5zm-6 7l-4-5h3V8h2v3h3l-4 5z"
                  />
                </g>
              </svg>
              <span
                class="woocommerce-table__download-button__label"
              >
                Download
              </span>
            </button>
          </div>
          <div
            class="woocommerce-ellipsis-menu"
          >
            <div
              class="components-dropdown"
              tabindex="-1"
            >
              <button
                aria-expanded="false"
                class="components-button woocommerce-ellipsis-menu__toggle"
                title="Choose which values to display"
                type="button"
              >
                <svg
                  class="gridicon gridicons-ellipsis"
                  height="24"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g>
                    <path
                      d="M7 12a2 2 0 11-4.001-.001A2 2 0 017 12zm12-2a2 2 0 10.001 4.001A2 2 0 0019 10zm-7 0a2 2 0 10.001 4.001A2 2 0 0012 10z"
                    />
                  </g>
                </svg>
              </button>
            </div>
          </div>
        </div>
        <div
          class="components-card__body components-card-body css-1i4jx7i-View-Body-borderRadius em57xhy0"
          data-wp-c16t="true"
          data-wp-component="CardBody"
        >
          <div
            aria-labelledby="caption-1"
            class="woocommerce-table__table"
            role="group"
          >
            <table>
              <caption
                class="woocommerce-table__caption screen-reader-text"
                id="caption-1"
              >
                Card readers
              </caption>
              <tbody>
                <tr>
                  <th
                    class="woocommerce-table__header is-left-aligned"
                    role="columnheader"
                    scope="col"
                  >
                    <span
                      aria-hidden="false"
                    >
                      Reader id
                    </span>
                  </th>
                  <th
                    class="woocommerce-table__header is-left-aligned"
                    role="columnheader"
                    scope="col"
                  >
                    <span
                      aria-hidden="false"
                    >
                      Status
                    </span>
                  </th>
                  <th
                    class="woocommerce-table__header is-left-aligned"
                    role="columnheader"
                    scope="col"
                  >
                    <span
                      aria-hidden="false"
                    >
                      Transactions
                    </span>
                  </th>
                  <th
                    class="woocommerce-table__header is-left-aligned"
                    role="columnheader"
                    scope="col"
                  >
                    <span
                      aria-hidden="false"
                    >
                      Fee
                    </span>
                  </th>
                </tr>
                <tr>
                  <th
                    class="woocommerce-table__item is-left-aligned"
                    scope="row"
                  >
                    1
                  </th>
                  <td
                    class="woocommerce-table__item is-left-aligned"
                  >
                    active
                  </td>
                  <td
                    class="woocommerce-table__item is-left-aligned"
                  >
                    3
                  </td>
                  <td
                    class="woocommerce-table__item is-left-aligned"
                  >
                    $3.00
                  </td>
                </tr>
                <tr>
                  <th
                    class="woocommerce-table__item is-left-aligned"
                    scope="row"
                  >
                    2
                  </th>
                  <td
                    class="woocommerce-table__item is-left-aligned"
                  >
                    inactive
                  </td>
                  <td
                    class="woocommerce-table__item is-left-aligned"
                  >
                    1
                  </td>
                  <td
                    class="woocommerce-table__item is-left-aligned"
                  >
                    $0.00
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
        <div
          class="components-flex components-card__footer components-card-footer css-142znvl-View-Flex-sx-Base-sx-Items-ItemsRow-Footer-borderRadius-borderColor-medium em57xhy0"
          data-wp-c16t="true"
          data-wp-component="CardFooter"
        />
      </div>
      <div
        aria-hidden="true"
        class="components-elevation css-91yjwm-View-Elevation-sx-Base-elevationClassName em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Elevation"
      />
      <div
        aria-hidden="true"
        class="components-elevation css-91yjwm-View-Elevation-sx-Base-elevationClassName em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Elevation"
      />
    </div>
  </div>
</div>
`;

exports[`RenderPaymentCardReaderChargeDetails renders without reader charges 1`] = `
<div>
  <div
    class="wcpay-payment-details woocommerce-payments-page"
    style="max-width: 1032px;"
  >
    <div
      class="components-surface components-card woocommerce-table transactions-list woocommerce-report-table has-search has-actions has-menu css-nsno0f-View-Surface-getBorders-primary-Card-rounded em57xhy0"
      data-wp-c16t="true"
      data-wp-component="Card"
    >
      <div
        class="css-mgwsf4-View-Content em57xhy0"
      >
        <div
          class="components-flex components-card__header components-card-header css-1g5oj2q-View-Flex-sx-Base-sx-Items-ItemsRow-Header-borderRadius-borderColor-medium em57xhy0"
          data-wp-c16t="true"
          data-wp-component="CardHeader"
        >
          <h2
            class="components-truncate components-text css-1s1vggh-View-Text-sx-Base em57xhy0"
            data-wp-c16t="true"
            data-wp-component="Text"
          >
            Card readers
          </h2>
          <div
            class="woocommerce-table__actions"
          />
          <div
            class="woocommerce-ellipsis-menu"
          >
            <div
              class="components-dropdown"
              tabindex="-1"
            >
              <button
                aria-expanded="false"
                class="components-button woocommerce-ellipsis-menu__toggle"
                title="Choose which values to display"
                type="button"
              >
                <svg
                  class="gridicon gridicons-ellipsis"
                  height="24"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g>
                    <path
                      d="M7 12a2 2 0 11-4.001-.001A2 2 0 017 12zm12-2a2 2 0 10.001 4.001A2 2 0 0019 10zm-7 0a2 2 0 10.001 4.001A2 2 0 0012 10z"
                    />
                  </g>
                </svg>
              </button>
            </div>
          </div>
        </div>
        <div
          class="components-card__body components-card-body css-1i4jx7i-View-Body-borderRadius em57xhy0"
          data-wp-c16t="true"
          data-wp-component="CardBody"
        >
          <div
            aria-labelledby="caption-0"
            class="woocommerce-table__table"
            role="group"
          >
            <table>
              <caption
                class="woocommerce-table__caption screen-reader-text"
                id="caption-0"
              >
                Card readers
              </caption>
              <tbody>
                <tr>
                  <th
                    class="woocommerce-table__header is-left-aligned"
                    role="columnheader"
                    scope="col"
                  >
                    <span
                      aria-hidden="false"
                    >
                      Reader id
                    </span>
                  </th>
                  <th
                    class="woocommerce-table__header is-left-aligned"
                    role="columnheader"
                    scope="col"
                  >
                    <span
                      aria-hidden="false"
                    >
                      Status
                    </span>
                  </th>
                  <th
                    class="woocommerce-table__header is-left-aligned"
                    role="columnheader"
                    scope="col"
                  >
                    <span
                      aria-hidden="false"
                    >
                      Transactions
                    </span>
                  </th>
                  <th
                    class="woocommerce-table__header is-left-aligned"
                    role="columnheader"
                    scope="col"
                  >
                    <span
                      aria-hidden="false"
                    >
                      Fee
                    </span>
                  </th>
                </tr>
                <tr>
                  <td
                    class="woocommerce-table__empty-item"
                    colspan="4"
                  >
                    No data to display
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
        <div
          class="components-flex components-card__footer components-card-footer css-142znvl-View-Flex-sx-Base-sx-Items-ItemsRow-Footer-borderRadius-borderColor-medium em57xhy0"
          data-wp-c16t="true"
          data-wp-component="CardFooter"
        />
      </div>
      <div
        aria-hidden="true"
        class="components-elevation css-91yjwm-View-Elevation-sx-Base-elevationClassName em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Elevation"
      />
      <div
        aria-hidden="true"
        class="components-elevation css-91yjwm-View-Elevation-sx-Base-elevationClassName em57xhy0"
        data-wp-c16t="true"
        data-wp-component="Elevation"
      />
    </div>
  </div>
</div>
`;
